compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2015,6 +2025,152 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                } else if found_refs.len() > expected_refs.len() {
+                    if found_refs[..found_refs.len() - expected_refs.len()]
+                        .iter()
+                        .zip(expected_refs.iter())
+                        .any(|(e, f)| e != f)
+                    {
+                        // The refs have different mutability.
+                        format!(
+                            "{}*{name}",
+                            found_refs[..found_refs.len() - expected_refs.len()]
+                                .iter()
+                                .map(|mutbl| format!("&{}", mutbl.prefix_str()))
+                                .collect::<Vec<_>>()
+                                .join(""),
+                        )
+                    } else {
+                        format!(
+                            "{}{name}",
+                            found_refs[..found_refs.len() - expected_refs.len()]
+                                .iter()
+                                .map(|mutbl| format!("&{}", mutbl.prefix_str()))
+                                .collect::<Vec<_>>()
+                                .join(""),
+                        )
+                    }
+                } else if expected_refs.len() > found_refs.len() {
+                    format!(
+                        "{}{name}",
+                        (0..(expected_refs.len() - found_refs.len()))
+                            .map(|_| "*")
+                            .collect::<Vec<_>>()
+                            .join(""),
+                    )
+                } else {
+                    format!(
+                        "{}{name}",
+                        found_refs
+                            .iter()
+                            .map(|mutbl| format!("&{}", mutbl.prefix_str()))
+                            .chain(found_refs.iter().map(|_| "*".to_string()))
+                            .collect::<Vec<_>>()
+                            .join(""),
+                    )
+                }
+            } else {
+                format!("/* {found} */")
+            }
+        };
+        let (closure_names, call_names): (Vec<_>, Vec<_>) =
+            if both_tys.clone().all(|(expected, found)| {
+                let (expected_ty, _) = get_deref_type_and_refs(expected);
+                let (found_ty, _) = get_deref_type_and_refs(found);
+                self.can_eq(param_env, found_ty, expected_ty)
+            }) && !expected_inputs.is_empty()
+                && expected_inputs.len() == found_inputs.len()
+                && let hir::QPath::Resolved(_, path) = path
+                && let hir::def::Res::Def(_, fn_def_id) = path.res
+                && let Some(node) = self.tcx.hir().get_if_local(fn_def_id)
+                && let Some(body_id) = node.body_id()
+            {
+                let closure = self
+                    .tcx
+                    .hir()
+                    .body_param_names(body_id)
```

Comment:
`query fn_arg_names` looks like a better version of this function, it works in more cases including functions from other crates.